test/app-luatest/uri_test.lua

```diff
@@ -0,0 +1,324 @@
+        "%F9", "%FA", "%FB", "%FC", "%FD", "%FE", "%FF",
+    },
+}
+
+local build_extended_ascii_str = function()
+    local str = ""
+    for c = 0, 255 do
+        str = str .. string.char(c)
+    end
+    return str
+end
+
+g_opts.test_uri_escape = function(cg)
+    local opts_name = cg.params.name
+    local opts = uri[opts_name]
+    -- Traverse an Extended ASCII table.
+    for c = 0, 255 do
+        local ch = string.char(c)
+        local expected = escaped_bytes[opts_name][c + 1]
+        t.assert_equals(uri.escape(ch, opts), expected)
+    end
+end
+
+g_opts.test_uri_unescape = function(cg)
+    local opts_name = cg.params.name
+    local opts = uri[opts_name]
+    -- Traverse an Extended ASCII table.
+    for c = 0, 255 do
+        local ch = string.char(c)
+        local actual = uri.unescape(escaped_bytes[opts_name][c + 1], opts)
+        t.assert_equals(actual, ch)
+    end
+end
+
+g.test_uri_unescape_icase= function(_)
+    local escaped_bytes_lowercase = "%1a%1b%1c%1d%1e%1f"
+    local escaped_bytes_uppercase = "%1A%1B%1C%1D%1E%1F"
+    local unescaped_lowercase = uri.unescape(escaped_bytes_lowercase)
+    local unescaped_uppercase = uri.unescape(escaped_bytes_uppercase)
+    t.assert_equals("\26\27\28\29\30\31", unescaped_lowercase)
+    t.assert_equals(unescaped_uppercase, unescaped_lowercase)
+end
+
+g.test_uri_unreserved_type = function(_)
+    t.assert_not_equals(uri.unreserved, nil)
+    t.assert_type(uri.unreserved, "function")
+    t.assert_type(uri.unreserved("string"), "cdata")
+end
+
+g.test_uri_unreserved = function(_)
+    local opts_pattern = {
+        unreserved = uri.unreserved("a-z"),
+    }
+    local opts_string = {
+        unreserved = uri.unreserved("abcdefghijklmnopqrstuvwxyz"),
+    }
+    local str = "abcdefghijklmnopqrstuvwxyzВК"
+    t.assert_equals(uri.escape(str, opts_string),
+                    "abcdefghijklmnopqrstuvwxyz%D0%92%D0%9A")
+    t.assert_equals(uri.escape(str, opts_pattern),
+                    "abcdefghijklmnopqrstuvwxyz%D0%92%D0%9A")
+end
+
+g_opts.test_uri_opts = function(cg)
+    local opts = uri[cg.params.name]
+    t.assert_not_equals(opts, nil)
+    t.assert_type(opts, "table")
+    t.assert_not_equals(opts.unreserved, nil)
+    t.assert_not_equals(opts.plus, nil)
+end
+
+g.test_uri_escape_RFC3986_is_default = function(_)
+    local str = build_extended_ascii_str()
+    local escaped_default_opts = uri.escape(str)
+    local escaped_RFC3986_opts = uri.escape(str, uri.RFC3986)
+    t.assert_equals(escaped_default_opts, escaped_RFC3986_opts)
+end
+
+g.test_uri_unescape_RFC3986_is_default = function(_)
+    local str = build_extended_ascii_str()
+    local escaped_RFC3986_opts = uri.escape(str, uri.RFC3986)
+    local unescaped_RFC3986_opts = uri.unescape(escaped_RFC3986_opts,
+                                                uri.RFC3986)
+    local unescaped_default_opts = uri.unescape(escaped_RFC3986_opts)
+    t.assert_equals(unescaped_RFC3986_opts, unescaped_default_opts)
+end
```

Comment:
The source sample (`escaped_RFC3986_opts`) has no `+` (it has `%2B`, but not `+`), so the test case will not differentiate `uri.unescape(<...>, {plus = false})` and `uri.unescape(<...>, {plus = true})`.

Reply:
1. `str` already has `+`.
2. We need it in `escaped_RFC3986_opts` (in the second test case), not in `str`.